src/isomorphic/ReactDebugTool.js

```diff
@@ -0,0 +1,88 @@
+/**
+ * Copyright 2016-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @providesModule ReactDebugTool
+ */
+
+'use strict';
+
+var ExecutionEnvironment = require('ExecutionEnvironment');
+var ReactPerfDevtool = require('ReactPerfDevtool');
+
+var warning = require('warning');
+
+var eventHandlers = [];
+var handlerDoesThrowForEvent = {};
+
+function emitEvent(handlerFunctionName, arg1, arg2, arg3, arg4, arg5) {
+  if (__DEV__) {
+    eventHandlers.forEach(function(handler) {
+      try {
+        if (handler[handlerFunctionName]) {
+          handler[handlerFunctionName](arg1, arg2, arg3, arg4, arg5);
+        }
+      } catch (e) {
+        warning(
+          !handlerDoesThrowForEvent[handlerFunctionName],
+          'exception thrown by devtool while handling %s: %s',
+          handlerFunctionName,
+          e.message
+        );
+        handlerDoesThrowForEvent[handlerFunctionName] = true;
+      }
+    });
+  }
+}
+
+var ReactDebugTool = {
+  addDevtool(devtool) {
+    eventHandlers.push(devtool);
+  },
+  removeDevtool(devtool) {
+    for (var i = 0; i < eventHandlers.length; i++) {
+      if (eventHandlers[i] === devtool) {
+        eventHandlers.splice(i, 1);
+        i--;
+      }
+    }
+  },
+  onBeginFlushUpdates() {
+    emitEvent('onBeginFlushUpdates');
+  },
+  onStartTimer(internalInstance, timerType) {
+    emitEvent('onStartTimer', internalInstance, timerType);
+  },
+  onEndFlushUpdates() {
+    emitEvent('onEndFlushUpdates');
+  },
+  onStopTimer(internalInstance, timerType) {
```

Comment:
>Would the work stack be completely divorced from the parent hierarchy?

Not completely but moving towards divorcing it is the plan, *especially* in perf-critical situations like infinite lists. We want to reconcile what’s on the screen with a priority so visible children will receive updates faster than other children and/or parents.

In addition we *already* do some pretty important work asynchronously—for example, we run `componentDidMount` outside the mounting work stack. Current ReactPerf completely miss it for this reason, but this version doesn’t.

I think we can still draw flame charts with the new approach. They could use the parent hierarchy for the stack levels, and reflect the situation more precisely by separating the actionable parts from the parts imposed by React itself.

That said I don’t have a vast experience building profiler tools so I may be off here.